eureka/S4_generate_lightcurves/clipping.py

```diff
@@ -0,0 +1,92 @@
+import numpy as np
+from astropy.convolution import Box1DKernel, convolve
+from astropy.stats import sigma_clip
+ 
+def clip_outliers(data, log, wavelength, sigma=10, box_width=5, maxiters=5, fill_value='boxcar', verbose=False):
+    '''Find outliers in 1D time series.
+  
+    Be careful when using this function on a time-series with known astrophysical variations. The variable
+    box_width should be set to be significantly smaller than any astrophysical variation timescales otherwise
+    these signals may be clipped.
+   
+    Parameters
+    ----------
+    data: ndarray (1D, float)
+        The input array in which to identify outliers
+    log: logedit.Logedit
+        The open log in which notes from this step can be added.
+    wavelength: float
+        The wavelength currently under consideration.
+    sigma: float
+        The number of sigmas a point must be from the rolling mean to be considered an outlier
+    box_width: int
+        The width of the box-car filter (used to calculated the rolling median) in units of number of data points
+    maxiters: int
+        The number of iterations of sigma clipping that should be performed.
+    fill_value: string or float
+        Either the string 'boxcar' to replace data with the mean from the box-car filter or a constant float-type fill value.
+  
+    Returns
+    -------
+    data:   ndarray (1D, boolean)
+        An array with the same dimensions as the input array with outliers replaced with fill_value.
+  
+    Notes
+    -----
+    History:
+  
+    - Jan 29-31, 2022 Taylor Bell
+        Initial version, added logging
+    '''
+    kernel = Box1DKernel(box_width)
+    # Compute the moving mean
+    smoothed_data = convolve(data, kernel, boundary='extend')
+    # Compare data to the moving mean (to remove astrophysical signals)
+    residuals = data-smoothed_data
+    # Sigma clip residuals to find bad points in data
+    residuals = sigma_clip(residuals, sigma=sigma, maxiters=maxiters)
+    outliers = np.ma.getmaskarray(residuals)
+  
+    if np.any(outliers) and verbose:
+        log.writelog('Identified {} outliers for wavelength {}'.format(np.sum(outliers), wavelength))
+  
+    # Replace clipped data
+    if fill_value=='boxcar':
```

Comment:
I don't like the idea of replacing outliers with mean values. This can mess with systematic-fitting routines in Stage 5 and lead to poor fits with outliers that normally would have been corrected. The better solution is to flag outliers as bad and leave the data values unchanged.

Reply:
Yeah, initially I wanted to do something like setting the values to NaN (or using numpy masked arrays, but they can use up a lot of RAM), but I wasn't sure how this would impact later stages. I look into this more

Reply:
Masked arrays should be fine at this stage. Using a boolean array of good/bad points also works.

Reply:
Alright, I haven't tested that masked arrays from S4 properly get passed to S5, but the new code should work. I'll test it tomorrow

Reply:
Alright, masked arrays work well in Stage 4 and 5, so marking as resolved. I kept the replacement code in place in case others want to do something similar in the future, but the code defaults to masking instead now.

Reply:
Oops, I was still using the replacement option rather than masking, and masking doesn't seem to work right now. Looking into it more